# Machine Learning Engineer Nanodegree
## Supervised Learning
## Project: Building a Student Intervention System

Welcome to the second project of the Machine Learning Engineer Nanodegree! In this notebook, some template code has already been provided for you, and it will be your job to implement the additional functionality necessary to successfully complete this project. Sections that begin with **'Implementation'** in the header indicate that the following block of code will require additional functionality which you must provide. Instructions will be provided for each section and the specifics of the implementation are marked in the code block with a `'TODO'` statement. Please be sure to read the instructions carefully!

In addition to implementing code, there will be questions that you must answer which relate to the project and your implementation. Each section where you will answer a question is preceded by a **'Question X'** header. Carefully read each question and provide thorough answers in the following text boxes that begin with **'Answer:'**. Your project submission will be evaluated based on your answers to each of the questions and the implementation you provide.  

>**Note:** Code and Markdown cells can be executed using the **Shift + Enter** keyboard shortcut. In addition, Markdown cells can be edited by typically double-clicking the cell to enter edit mode.

### Question 1 - Classification vs. Regression
*Your goal for this project is to identify students who might need early intervention before they fail to graduate. Which type of supervised learning problem is this, classification or regression? Why?*

**Answer: ** It could be both, regression would predict grades, and classifications just if is going to fail to graduate or not. But in a more strict sense, should be classification. Students who need help and who doesnt, also the dataset doesnt have grades.

## Exploring the Data
Run the code cell below to load necessary Python libraries and load the student data. Note that the last column from this dataset, `'passed'`, will be our target label (whether the student graduated or didn't graduate). All other columns are features about each student.

In [1]:
# Import libraries
import numpy as np
import pandas as pd
from time import time
from sklearn.metrics import f1_score

# Read student data
student_data = pd.read_csv("student-data.csv")
print "Student data read successfully!"


Student data read successfully!


### Implementation: Data Exploration
Let's begin by investigating the dataset to determine how many students we have information on, and learn about the graduation rate among these students. In the code cell below, you will need to compute the following:
- The total number of students, `n_students`.
- The total number of features for each student, `n_features`.
- The number of those students who passed, `n_passed`.
- The number of those students who failed, `n_failed`.
- The graduation rate of the class, `grad_rate`, in percent (%).


In [2]:
# TODO: Calculate number of students
n_students = student_data.shape[0]

# TODO: Calculate number of features
#without last column (passed or not)
n_features = student_data.shape[1] - 1 #30

# TODO: Calculate passing students
n_passed = student_data[student_data.passed == 'yes'].shape[0]

# TODO: Calculate failing students
n_failed = n_students - n_passed

# TODO: Calculate graduation rate
grad_rate = float(n_passed*100)/float(n_students)

# Print the results
print "Total number of students: {}".format(n_students)
print "Number of features: {}".format(n_features)
print "Number of students who passed: {}".format(n_passed)
print "Number of students who failed: {}".format(n_failed)
print "Graduation rate of the class: {:.2f}%".format(grad_rate)

Total number of students: 395
Number of features: 30
Number of students who passed: 265
Number of students who failed: 130
Graduation rate of the class: 67.09%


## Preparing the Data
In this section, we will prepare the data for modeling, training and testing.

### Identify feature and target columns
It is often the case that the data you obtain contains non-numeric features. This can be a problem, as most machine learning algorithms expect numeric data to perform computations with.

Run the code cell below to separate the student data into feature and target columns to see if any features are non-numeric.

In [3]:
# Extract feature columns
feature_cols = list(student_data.columns[:-1])

# Extract target column 'passed'
target_col = student_data.columns[-1]

# Show the list of columns
print "Feature columns:\n{}".format(feature_cols)
print "\nTarget column: {}".format(target_col)

# Separate the data into feature data and target data (X_all and y_all, respectively)
X_all = student_data[feature_cols]
y_all = student_data[target_col]

# Show the feature information by printing the first five rows
print "\nFeature values:"
print X_all.head()


Feature columns:
['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu', 'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']

Target column: passed

Feature values:
  school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

    ...    higher internet  romantic  famrel  freetime goout Dalc Walc health  \
0   ...       yes       no        no       4         3     4    1    1      3   
1   ...       

### Preprocess Feature Columns

As you can see, there are several non-numeric columns that need to be converted! Many of them are simply `yes`/`no`, e.g. `internet`. These can be reasonably converted into `1`/`0` (binary) values.

Other columns, like `Mjob` and `Fjob`, have more than two values, and are known as _categorical variables_. The recommended way to handle such a column is to create as many columns as possible values (e.g. `Fjob_teacher`, `Fjob_other`, `Fjob_services`, etc.), and assign a `1` to one of them and `0` to all others.

These generated columns are sometimes called _dummy variables_, and we will use the [`pandas.get_dummies()`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html?highlight=get_dummies#pandas.get_dummies) function to perform this transformation. Run the code cell below to perform the preprocessing routine discussed in this section.

In [4]:
def preprocess_features(X):
    ''' Preprocesses the student data and converts non-numeric binary variables into
        binary (0/1) variables. Converts categorical variables into dummy variables. '''
    
    # Initialize new output DataFrame
    output = pd.DataFrame(index = X.index)

    # Investigate each feature column for the data
    for col, col_data in X.iteritems():
        
        # If data type is non-numeric, replace all yes/no values with 1/0
        if col_data.dtype == object:
            col_data = col_data.replace(['yes', 'no'], [1, 0])

        # If data type is categorical, convert to dummy variables
        if col_data.dtype == object:
            # Example: 'school' => 'school_GP' and 'school_MS'
            col_data = pd.get_dummies(col_data, prefix = col)  
        
        # Collect the revised columns
        output = output.join(col_data)
    
    return output

X_all = preprocess_features(X_all)
print "Processed feature columns ({} total features):\n{}".format(len(X_all.columns), list(X_all.columns))

Processed feature columns (48 total features):
['school_GP', 'school_MS', 'sex_F', 'sex_M', 'age', 'address_R', 'address_U', 'famsize_GT3', 'famsize_LE3', 'Pstatus_A', 'Pstatus_T', 'Medu', 'Fedu', 'Mjob_at_home', 'Mjob_health', 'Mjob_other', 'Mjob_services', 'Mjob_teacher', 'Fjob_at_home', 'Fjob_health', 'Fjob_other', 'Fjob_services', 'Fjob_teacher', 'reason_course', 'reason_home', 'reason_other', 'reason_reputation', 'guardian_father', 'guardian_mother', 'guardian_other', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']


### Implementation: Training and Testing Data Split
So far, we have converted all _categorical_ features into numeric values. For the next step, we split the data (both features and corresponding labels) into training and test sets. In the following code cell below, you will need to implement the following:
- Randomly shuffle and split the data (`X_all`, `y_all`) into training and testing subsets.
  - Use 300 training points (approximately 75%) and 95 testing points (approximately 25%).
  - Set a `random_state` for the function(s) you use, if provided.
  - Store the results in `X_train`, `X_test`, `y_train`, and `y_test`.

In [5]:
# TODO: Import any additional functionality you may need here
from sklearn.model_selection import train_test_split
# TODO: Set the number of training points
num_train = 95

# Set the number of testing points
num_test = X_all.shape[0] - num_train

# TODO: Shuffle and split the dataset into the number of training and testing points above
X_train, X_test, y_train, y_test = train_test_split(X_all, y_all, train_size=num_test, random_state=24)

# Show the results of the split
print "Training set has {} samples.".format(X_train.shape[0])
print "Testing set has {} samples.".format(X_test.shape[0])


Training set has 300 samples.
Testing set has 95 samples.


## Training and Evaluating Models
In this section, you will choose 3 supervised learning models that are appropriate for this problem and available in `scikit-learn`. You will first discuss the reasoning behind choosing these three models by considering what you know about the data and each model's strengths and weaknesses. You will then fit the model to varying sizes of training data (100 data points, 200 data points, and 300 data points) and measure the F<sub>1</sub> score. You will need to produce three tables (one for each model) that shows the training set size, training time, prediction time, F<sub>1</sub> score on the training set, and F<sub>1</sub> score on the testing set.

**The following supervised learning models are currently available in** [`scikit-learn`](http://scikit-learn.org/stable/supervised_learning.html) **that you may choose from:**
- Gaussian Naive Bayes (GaussianNB)
- Decision Trees
- Ensemble Methods (Bagging, AdaBoost, Random Forest, Gradient Boosting)
- K-Nearest Neighbors (KNeighbors)
- Stochastic Gradient Descent (SGDC)
- Support Vector Machines (SVM)
- Logistic Regression


### Question 2 - Model Application
*List three supervised learning models that are appropriate for this problem. For each model chosen*
- Describe one real-world application in industry where the model can be applied. *(You may need to do a small bit of research for this — give references!)* 
- What are the strengths of the model; when does it perform well? 
- What are the weaknesses of the model; when does it perform poorly?
- What makes this model a good candidate for the problem, given what you know about the data?

**Answer: ** SVM, Decision Trees, GaussianNB

 - **SVM:**
     - **Finalcials:** 
         Used in credit rankings on the markets, to analyze riskness.
         - ref: http://www.svms.org/finance/Huang-etal2004.pdf
     - **Strenghts:**
         - Works well with high dimensional data.
         - Capture complex relationships between data.
     - **Weakness:**
         - Is not good with high dimensional features and low amount of data to train.
         - Is not a good with multiple classes.
     - Is a good model to use because:
         - We are dealing with 2 classes (pass, fail)
         - We are using a lot of features.
 - **Decision Trees:**
     - **Predicting Students Final GPA Using Decision Trees**
         Desion trees are used deeply in education.
         - ref:  http://www.ijiet.org/vol6/745-IT205.pdf
     - **Strenghts:**
         - Does not require data normalized.
         - It makes nonlinear relationships.
     - **Weakness:**
         - Smooth boundaries.
         - Instability, with small perturbations on data sets the model could change. (noise)
     - Is a good model to use because:
         - We are not using linear data. And is a classification problem.

 - **Gaussian Naive Bayes:**
    - **Diagnosing insurance claim fraud**
        -ref: http://ieeexplore.ieee.org/document/1277822/
    - **Strenghts:**
        - Simply model.
        - Works with less data for training in comparison with others algorithms.
    - **Weakness:**
        - It doesn't allow for rich hypotheses.
    - Is a good model to use because:
        - Generalize well, and it could also works with missing data.
        


### Setup
Run the code cell below to initialize three helper functions which you can use for training and testing the three supervised learning models you've chosen above. The functions are as follows:
- `train_classifier` - takes as input a classifier and training data and fits the classifier to the data.
- `predict_labels` - takes as input a fit classifier, features, and a target labeling and makes predictions using the F<sub>1</sub> score.
- `train_predict` - takes as input a classifier, and the training and testing data, and performs `train_clasifier` and `predict_labels`.
 - This function will report the F<sub>1</sub> score for both the training and testing data separately.

In [6]:
def train_classifier(clf, X_train, y_train):
    ''' Fits a classifier to the training data. '''
    
    # Start the clock, train the classifier, then stop the clock
    start = time()
    clf.fit(X_train, y_train)
    end = time()
    
    # Print the results
    print "Trained model in {:.4f} seconds".format(end - start)

    
def predict_labels(clf, features, target):
    ''' Makes predictions using a fit classifier based on F1 score. '''
    
    # Start the clock, make predictions, then stop the clock
    start = time()
    y_pred = clf.predict(features)
    end = time()
    
    # Print and return results
    print "Made predictions in {:.4f} seconds.".format(end - start)
    return f1_score(target.values, y_pred, pos_label='yes')


def train_predict(clf, X_train, y_train, X_test, y_test):
    ''' Train and predict using a classifer based on F1 score. '''
    
    # Indicate the classifier and the training set size
    print "Training a {} using a training set size of {}. . .".format(clf.__class__.__name__, len(X_train))
    
    # Train the classifier
    train_classifier(clf, X_train, y_train)
    
    # Print the results of prediction for both training and testing
    print "F1 score for training set: {:.4f}.".format(predict_labels(clf, X_train, y_train))
    print "F1 score for test set: {:.4f}.".format(predict_labels(clf, X_test, y_test))

### Implementation: Model Performance Metrics
With the predefined functions above, you will now import the three supervised learning models of your choice and run the `train_predict` function for each one. Remember that you will need to train and predict on each classifier for three different training set sizes: 100, 200, and 300. Hence, you should expect to have 9 different outputs below — 3 for each model using the varying training set sizes. In the following code cell, you will need to implement the following:
- Import the three supervised learning models you've discussed in the previous section.
- Initialize the three models and store them in `clf_A`, `clf_B`, and `clf_C`.
 - Use a `random_state` for each model you use, if provided.
 - **Note:** Use the default settings for each model — you will tune one specific model in a later section.
- Create the different training set sizes to be used to train each model.
 - *Do not reshuffle and resplit the data! The new training points should be drawn from `X_train` and `y_train`.*
- Fit each model with each training set size and make predictions on the test set (9 in total).  
**Note:** Three tables are provided after the following code cell which can be used to store your results.

In [7]:
# TODO: Import the three supervised learning models from sklearn
# from sklearn import model_A
# from sklearn import model_B
# from skearln import model_C

from sklearn import svm
from sklearn import tree
from sklearn import naive_bayes

# TODO: Initialize the three models
clf_A = svm.SVC(random_state=10)
clf_B = tree.tree.DecisionTreeClassifier(random_state=20)
clf_C = naive_bayes.GaussianNB()

from sklearn.model_selection import train_test_split

# TODO: Set up the training set sizes
train_100 = [X_train[100:], X_test, y_train[100:], y_test]
train_200 = [X_train[200:], X_test, y_train[200:], y_test]
train_300 = [X_train, X_test, y_train, y_test]

print "####  #####"

train_predict(clf_A, train_100[0], train_100[2], train_100[1], train_100[3])
train_predict(clf_A, train_200[0], train_200[2], train_200[1], train_200[3])
train_predict(clf_A, train_300[0], train_300[2], train_300[1], train_300[3])

print "####  #####"


train_predict(clf_B, train_100[0], train_100[2], train_100[1], train_100[3])
train_predict(clf_B, train_200[0], train_200[2], train_200[1], train_200[3])
train_predict(clf_B, train_300[0], train_300[2], train_300[1], train_300[3])

print "####  #####"
train_predict(clf_C, train_100[0], train_100[2], train_100[1], train_100[3])
train_predict(clf_C, train_200[0], train_200[2], train_200[1], train_200[3])
train_predict(clf_C, train_300[0], train_300[2], train_300[1], train_300[3])
print "####  #####"


# TODO: Execute the 'train_predict' function for each classifier and each training set size
# train_predict(clf, X_train, y_train, X_test, y_test)

####  #####
Training a SVC using a training set size of 200. . .
Trained model in 0.0055 seconds
Made predictions in 0.0027 seconds.
F1 score for training set: 0.8716.
Made predictions in 0.0012 seconds.
F1 score for test set: 0.8219.
Training a SVC using a training set size of 100. . .
Trained model in 0.0012 seconds
Made predictions in 0.0008 seconds.
F1 score for training set: 0.8537.
Made predictions in 0.0007 seconds.
F1 score for test set: 0.8250.
Training a SVC using a training set size of 300. . .
Trained model in 0.0060 seconds
Made predictions in 0.0044 seconds.
F1 score for training set: 0.8559.
Made predictions in 0.0018 seconds.
F1 score for test set: 0.8108.
####  #####
Training a DecisionTreeClassifier using a training set size of 200. . .
Trained model in 0.0035 seconds
Made predictions in 0.0006 seconds.
F1 score for training set: 1.0000.
Made predictions in 0.0004 seconds.
F1 score for test set: 0.7874.
Training a DecisionTreeClassifier using a training set size of 10

### Tabular Results
Edit the cell below to see how a table can be designed in [Markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#tables). You can record your results from above in the tables provided.

** Classifer 1 - SVM**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |        0.0013           |       0.0009           |    0.8481        |      0.8115     |
| 200               |        0.0033           |       0.0003           |    0.8875        |      0.7961     |
| 300               |        0.0060           |       0.0043           |    0.8559        |      0.8108     |

** Classifer 2 - Decision Trees**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |     0.0011              |        0.0003          |  1.0000          |       0.7018    |
| 200               |     0.0014              |        0.0003          |  1.0000          |       0.7212    |
| 300               |     0.0020              |        0.0002          |  1.0000          |     0.7852      |

** Classifer 3 - GaussianNB**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |    0.0007               |        0.0004          |        0.8308    |    0.6935       |
| 200               |    0.0008               |        0.0005          |        0.8000    |    0.7054       |
| 300               |    0.0013               |        0.0006          |        0.7704    |    0.7812       |

## Choosing the Best Model
In this final section, you will choose from the three supervised learning models the *best* model to use on the student data. You will then perform a grid search optimization for the model over the entire training set (`X_train` and `y_train`) by tuning at least one parameter to improve upon the untuned model's F<sub>1</sub> score. 

### Question 3 - Choosing the Best Model
*Based on the experiments you performed earlier, in one to two paragraphs, explain to the board of supervisors what single model you chose as the best model. Which model is generally the most appropriate based on the available data, limited resources, cost, and performance?*

**Answer: ** Based apparently on test data scores the most accurate model is provided by SVM on spite of that it seems to overfit data more than GaussianNB, but SVM could be tuned better and final scores may be greater than GaussainNB. 
Considering this is not a real time problem, therefore, costs of resources and performance are ignored, because the system could be training for a week long and its not going to be a mayor problem, we don't require a computing lab to build this model.

### Question 4 - Model in Layman's Terms
*In one to two paragraphs, explain to the board of directors in layman's terms how the final model chosen is supposed to work. Be sure that you are describing the major qualities of the model, such as how the model is trained and how the model makes a prediction. Avoid using advanced mathematical or technical jargon, such as describing equations or discussing the algorithm implementation.*

**Answer: ** 
We have a few characteristics related to some students that we think tell us about enough to discriminate between a student who will succeed or fail. The main idea to create a predictve model using these characteristics is figure out if a student will need help or not. To achieve this it is necesary to create a mathematical model which allow us to predict this situation: SVM. First at all we have to train to understand that there's two kind or two classes of students, ones who fails and ones who succeed. The SVM will create a line between them both categories, this line will separate characteristics about students, and using advanced optimization processes the algorithm finds the best line which divide both classes. This line is based on few equidistant features-points of both categories, this points are called *suport vectors*.

Some times the data is not able to be separated by a single line, because the features are together and there's no clue to separate them. To solve this issue we use the kernel trick, and the idea of this is to add a new dimension to the problem using some heuristic to diferentiate both categories.


### Implementation: Model Tuning
Fine tune the chosen model. Use grid search (`GridSearchCV`) with at least one important parameter tuned with at least 3 different values. You will need to use the entire training set for this. In the code cell below, you will need to implement the following:
- Import [`sklearn.grid_search.GridSearchCV`](http://scikit-learn.org/0.17/modules/generated/sklearn.grid_search.GridSearchCV.html) and [`sklearn.metrics.make_scorer`](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html).
- Create a dictionary of parameters you wish to tune for the chosen model.
 - Example: `parameters = {'parameter' : [list of values]}`.
- Initialize the classifier you've chosen and store it in `clf`.
- Create the F<sub>1</sub> scoring function using `make_scorer` and store it in `f1_scorer`.
 - Set the `pos_label` parameter to the correct value!
- Perform grid search on the classifier `clf` using `f1_scorer` as the scoring method, and store it in `grid_obj`.
- Fit the grid search object to the training data (`X_train`, `y_train`), and store it in `grid_obj`.

In [8]:
# TODO: Import 'GridSearchCV' and 'make_scorer'
from sklearn.metrics import make_scorer, f1_score
from sklearn.model_selection import GridSearchCV
from sklearn import svm, naive_bayes
# TODO: Create the parameters list you wish to tune
from sklearn.preprocessing import StandardScaler

def f1_scorer(y_true, y_pred, labels=None, pos_label=1, average='binary', sample_weight=None):
    return f1_score(y_true, y_pred, pos_label='yes')

clf = svm.SVC()

def custom_kernel(X,Y):
    
    return None
parameters = [{'kernel': ['rbf'], 'C': [1,10], 'gamma': [0.1, 0.01, 0.001, 0.0001]},
                    {'kernel': ['linear'], 'C': [1,10]}, 
              {'kernel': ['poly'], 'degree': [1,10], 'C': [1,10], 'coef0': [0.1,0.5,0.9,0]},
              {'kernel': ['sigmoid'], 'degree': [1,10], 'C': [1,10]}]

grid_obj = GridSearchCV(clf, parameters, scoring=make_scorer(f1_scorer))

#X_new = VarianceThreshold(threshold=0.1).fit_transform(X_all)
#X_new = StandardScaler().fit_transform(X_all)
#X_train, X_test, y_train, y_test = train_test_split(X_new, y_all, random_state=24)

grid_obj.fit(X_train, y_train)

clf = grid_obj.best_estimator_

# Report the final F1 score for training and testing after parameter tuning

print "Tuned model has a training F1 score of {:.4f}.".format(predict_labels(clf, X_train, y_train))
print "Tuned model has a testing F1 score of {:.4f}.".format(predict_labels(clf, X_test, y_test))
print grid_obj.best_params_


Made predictions in 0.0036 seconds.
Tuned model has a training F1 score of 0.8170.
Made predictions in 0.0015 seconds.
Tuned model has a testing F1 score of 0.8553.
{'kernel': 'rbf', 'C': 10, 'gamma': 0.001}


### Question 5 - Final F<sub>1</sub> Score
*What is the final model's F<sub>1</sub> score for training and testing? How does that score compare to the untuned model?*

**Answer: ** The scores are slightly better with the tunned version, probably it might be better give a chance to others algorithms like random forest or KNN, and another possibility is to define a better custom kernel function to the SVM model.

> **Note**: Once you have completed all of the code implementations and successfully answered each question above, you may finalize your work by exporting the iPython Notebook as an HTML document. You can do this by using the menu above and navigating to  
**File -> Download as -> HTML (.html)**. Include the finished document along with this notebook as your submission.